### ISSUE TITLE
Let completion spec request fresh pseudo terminal with full $PATH and env context

### ISSUE THREAD

**Issue by wilhelmklopp** [withfig/fig#225]
## Is your feature request related to a problem? Please describe.
I'm working on the [Django completion spec](https://github.com/withfig/autocomplete/issues/298)

Django has a bunch of built-in tooling that provides information about the available subcommands, options, arguments, etc. I'm trying to build the Fig completion spec on top of this existing tooling

To make use of this built-in tooling, we can run `python manage.py help` or `python manage.py help runserver` and Django provides us with all we need to know: options, arguments, descriptions, etc.

And herein lies the problem ☝️ When using `executeShellCommand`, the pseudo terminal that actually executes the command doesn't share the exact same context as the terminal that the user is typing in. The user's terminal might have an activated python virtual environment, or other environment variables set that are necessary for `python manage.py *` commands to work. 

Since the pseudo terminal is missing that context, running `python manage.py help` is likely to fail and our spec won't be able to retrieve the information it needs to populate the autocomplete suggestions

## Describe the solution you'd like
Somewhere within my spec, I'd like to have a way to specify that I would like a "fresh" pseudo terminal, which shares all of the context as the shell the user is typing in. It should have the same $PATH and the same environment variables as the user's shell. If running `python manage.py help` succeeds in the user's shell, it should also succeed in the pseudo terminal that's running `executeShellCommand("python manage.py help")`

In terms of the API from the spec author perspective, one idea could be to make this an option within `executeShellCommand` . So perhaps I could do something like `executeShellCommand("python manage.py help", {freshPseudoTerminal: true})` – just an idea though!

#### How should this work under the hood?
I definitely only have a very rough understanding of how fig works, but one approach could be to run `env` in the user's shell, and then re-export every entry from the `env` output in the pseudo terminal. That should set $PATH and also all other environment variables that might be required in order for `python manage.py help` to work.

![image](https://user-images.githubusercontent.com/7718702/123509912-29267700-d670-11eb-8ebc-7ddabc353be7.png)

My understanding based on this ☝️ is that fig could access `env` every time the prompt refreshes. So maybe the pseudo terminal env could be set based on the most recently stored `env` from when the prompt last refreshed in a given tty?

## Describe alternatives you've considered
I think the main alternative that exists to all of this is to build the Django completion spec statically, and don't try to generate it at runtime based on the output from `python manage.py help`. This has the downside that we wouldn't be able to show completions for custom user-defined management commands. It would also decouple the spec from django itself, which presumably increases the maintenance burden of the spec quite substantially over time.

cc @brendanfalk @mattschrage

**Comment by QuiiBz** [comment]
That's definitely a good point.

> Somewhere within my spec, I'd like to have a way to specify that I would like a "fresh" pseudo terminal

Do you mean that, for instance, `executeShellCommand` can have an option to specify running the command in a new, fresh shell? The downside of this is that creating a shell is very slow, especially if it relies on `generateSpec`, which can be called very often.

One other option can be - same as your idea of sharing $PATH - specifying in an option of `executeShellCommand` the `env` to use. The env would be applied when running the command, and then re-set back to the default env. This could avoid the very slow part of spawning pseudo-terminals.

**Comment by wilhelmklopp** [comment]
> Do you mean that, for instance, executeShellCommand can have an option to specify running the command in a new, fresh shell? The downside of this is that creating a shell is very slow, especially if it relies on generateSpec, which can be called very often.

yeah this was one option I was thinking about. Really appreciate the context that creating a new shell is very slow, that helps my mental model ❤️ 



Essentially, my workflow as a user looks like this:

1. I run `pipenv shell`. This sets $PATH to add the python virtual environment. it also reads `.env` and sets all those as environment variables. This is a prerequisite for the commands in step 2 to work 👇 

2. Then I run my `python manage.py` commands (ie `python manage.py test` or `python manage.py runserver`). Running these only works since I've already run `pipenv shell`

What I'm looking for is a way that `executeShellCommand` can run against a shell that is identical to the user's shell that exists _after_ step 1. So after `pipenv shell` has finished.

Very open to any solution that could make this workflow work, whether it's specifying something within the API of `excuteShellCommand` or somewhere else in the spec, or some other approach entirely.

---

**New idea:** 
What if instead of spawning a whole new shell, I could just ask for the $PATH in the pseudo terminal to be set to the same value as the one in the user's shell. Presumably that would be a lot faster than spawning a new shell? So something like: `executeShellCommand("python manage.py help", {useUserPATH: true})`

There are alternative ways of setting the `.env` variables (ie from within python using [python-dotenv](https://pypi.org/project/python-dotenv/)), so I'm doing that now – but I think there's no way of getting around $PATH. That just needs to be present in the pseudoterminal for all of this to work


> One other option can be - same as your idea of sharing $PATH - specifying in an option of executeShellCommand the env to use. The env would be applied when running the command, and then re-set back to the default env. This could avoid the very slow part of spawning pseudo-terminals.

ooo this sounds very interesting, say more? Specifically curious about how specifying the "the env to use" could work 👀

**Comment by QuiiBz** [comment]
> ooo this sounds very interesting, say more? Specifically curious about how specifying the "the env to use" could work

A bit like your new idea, that would be an option in `executeShellCommand` which have a list of env keys to "preserve" (in reality they would be added/updated in the shell running the command, then reset back to the previous env) from the current shell.

In your case, that would be something like:
```ts
executeShellCommand("python manage.py help", { preserveEnv: ["PATH"] })
```

This would allow for more customisations in the futur, to cover every env variables and not only $PATH.

**Comment by wilhelmklopp** [comment]
sweet, yeah that sounds great. 

Would be awesome also if you could specify `preserveEnv: ["*"]` to copy over **all** env vars.

Then I wouldn't even have to use `python-dotenv` anymore. But if that's significantly more work, I'd be super happy with just `PATH` as a start 🙌

**Comment by mattschrage** [comment]
This is something I am going to look into this week @wilhelmklopp!

**Comment by mschrage** [comment]
Linking to #468 #1187 #882
